### ISSUE TITLE
Async version of Callbacks::tls_sign_message()

### ISSUE THREAD

**Issue by rojkov** [randombit/botan#1856]
Is there a plan to make `tls_sign_message()` asynchronous?

I'd like to offload TLS handshakes to Intel QAT hardware, but in the sync mode the performance gain is rather insignificant.

**Comment by randombit** [comment]
I don't think it would help much to make that async because it is used to sign the server key exchange, which is usually the last message in the flight. So until the signature is generated the handshake progress is blocked anyway.

Maybe I am missing something where this would still be beneficial.  I don't know much about QAT. Maybe QAT can do fast batching signs, such that if 4-8 handshakes are happening concurrently QAT can do them faster than it was given the same number of sync operations? But in that case, `tls_sign_message` implementation could use QAT in async mode and then just block internally until QAT returns, giving QAT the option to do batching.

(In theory we could kick off the signature generation, return immediately without finishing the handshake processing, then resume after the signature is finished. That would require a fairly big restructuring of the code and I'd rather attempt that when we can move to C++20 which has a lot more library help for things like this.)

**Comment by rojkov** [comment]
Thank you for your prompt reply!

I was referring [this](https://commondatastorage.googleapis.com/chromium-boringssl-docs/ssl.h.html#ssl_private_key_method_st) sort of API.

My use case is to have TLS handshakes offloaded to QAT in [a custom Envoy filter](https://github.com/envoyproxy/envoy-filter-example) and not to use Envoy's built-in TLS context. It was a very easy task to implement a TCP filter using Botan that wraps Envoy connections in TLS, by the way!

QAT works in such a way that a process is supposed to write a chunk of data sufficient for a signing task into QAT's input ring buffer and proceed with handling other tasks right away. When the task is complete QAT notifies the process by putting the result into an output ring buffer. CPU doesn't do any computations besides writing/reading from the buffers.

In case of Envoy its workers handle many connections in a single event loop per hardware thread. If  `tls_sign_message` blocked then other connections would stall until the result reached the output buffer.

**Comment by randombit** [comment]
Thanks for explaining the use case. Supporting this means we have to interrupt the server handshake flow mid-way, return to the application, then come back to it later on after the signature has completed to finish sending the handshake flight. And, ideally, do all this in such a way as to not cause any major overhead or latency in the common case of someone using the default software method.

But in normal course of events (ie how applications using the API currently work) no IO will occur, because the client isn't going to reply until after the server has sent the hello done, which happens after the (signed) server key exchange. And IO is the main event which causes an application to call back into a `Channel` object. But until the control flow passes back to the `Channel`, there is no way to check the status of the signature computation... 

I'm not really sure how to support this. If this was something specific to QAT I'd probably say sorry, pass, but this exact same situation comes up with any external hardware, and I know there are companies using the TLS stack with signature keys in TPM and PKCS11 hardware, so it's probably of wider interest. Thanks for linking to the BoringSSL API. I didn't know BoringSSL supported this. That's actually very handy as I can study how they approach the problem. That will save me a lot of time and thinking ;)

Some good news/bad news. I had thought that `std::future::then` had finally made it in C++20 but it appears it is even missing there. Bad, because that was my best hope for a clean and simple implementation. Good news, there is no reason to hold off waiting for better library support, because it's not coming anytime soon.

Right now I am very busy with {work,family,other projects} so can't promise much about when exactly anything will happen with this.

**Comment by rojkov** [comment]
Yes, this offloading starts to make financial sense when there are hundreds of simultaneous new connections served by a single CPU core like in case of big enough CDNs.

Just thinking out loud... `tls_sign_message()` could create a side channel and register a callback for it in the application's event loop (e.g. in Envoy's case with [this API](https://github.com/envoyproxy/envoy/blob/v1.9.0/include/envoy/event/dispatcher.h#L92)). When QAT puts data to that channel the registered callback will be called from the loop and it will pass the control flow to the `Channel`...

I wish I had time to play with it :)

**Comment by randombit** [comment]
Some benchmarks of QAT in async vs sync mode, clearly async is required to get good performance: https://www.servethehome.com/intel-quickassist-technology-and-openssl-setup-insights-and-initial-benchmarks/
